grin/src/Transformations/ExtendedSyntax/Optimising/CaseCopyPropagation.hs

```diff
@@ -0,0 +1,81 @@
+{-# LANGUAGE LambdaCase #-}
+module Transformations.ExtendedSyntax.Optimising.CaseCopyPropagation where
+
+import Data.Map (Map)
+import Data.Functor.Foldable
+
+import qualified Data.Map as Map
+
+import Control.Monad.State
+
+import Grin.ExtendedSyntax.Grin
+import Transformations.ExtendedSyntax.Names
+import Transformations.ExtendedSyntax.Util (cataM)
+
+
+-- NOTE: ~ Maybe Tag
+data TagInfo = Unknown | Known Tag
+  deriving (Eq, Ord, Show)
+
+-- | Maps alt names to TagInfo
+type InfoTable = Map Name TagInfo
+
+-- NOTE: Case Copy Propagtion ~ Case Unboxing
+caseCopyPropagation :: Exp -> (Exp, ExpChanges)
+caseCopyPropagation e = rebindCases infoTable e where
+  infoTable = collectTagInfo e
+
+-- | Collects tag information about case alternatives.
+collectTagInfo :: Exp -> InfoTable
+collectTagInfo = flip execState mempty . cataM alg where
+
+  alg :: ExpF TagInfo -> State InfoTable TagInfo
+  alg = \case
```

Comment:
`pure . \case`

Reply:
Yep, it _seems_ that you can do this, but actually you can't due to the non-trivial `AltF` branch. I didn't notice it the first time either (there is a sneaky newline).